options/options.go

```diff
@@ -795,5 +803,245 @@ type EngineOptions struct {
+
+			// Deep copy the Signals map
+			for sigKey, sigVal := range ignoreConfig.Signals {
+				cloned.Ignore[key].Signals[sigKey] = sigVal
+			}
+		}
+	}
+
+	return cloned
+}
+
+// RunWithErrorHandling runs the given operation and handles any errors according to the configuration.
+func (opts *TerragruntOptions) RunWithErrorHandling(ctx context.Context, operation func() error) error {
+	if opts.Errors == nil {
+		return operation()
+	}
+
+	currentAttempt := 1
+
+	for {
+		err := operation()
+		if err == nil {
+			return nil
+		}
+
+		// Process the error through our error handling configuration
+		action, processErr := opts.Errors.ProcessError(err, currentAttempt)
+		if processErr != nil {
+			return fmt.Errorf("error processing error handling rules: %w", processErr)
+		}
+
+		if action == nil {
+			return err
+		}
+
+		if action.ShouldIgnore {
+			opts.Logger.Warnf("Ignoring error, reason: %s", action.IgnoreMessage)
+
+			// Handle ignore signals if any are configured
+			if len(action.IgnoreSignals) > 0 {
+				if err := opts.handleIgnoreSignals(action.IgnoreSignals); err != nil {
+					return err
+				}
+			}
+
+			return nil
+		}
+
+		if action.ShouldRetry {
+			opts.Logger.Warnf(
+				"Encountered retryable error: %s\nAttempt %d of %d. Waiting %d second(s) before retrying...",
+				action.RetryMessage,
+				currentAttempt,
+				action.RetryAttempts,
+				action.RetrySleepSecs,
+			)
+
+			// Sleep before retry
+			select {
+			case <-time.After(time.Duration(action.RetrySleepSecs) * time.Second):
+				// try again
+			case <-ctx.Done():
+				return errors.New(ctx.Err())
+			}
+
+			currentAttempt++
+
+			continue
+		}
+
+		return err
+	}
+}
+
+func (opts *TerragruntOptions) handleIgnoreSignals(signals map[string]interface{}) error {
+	workingDir := opts.WorkingDir
+	signalsFile := filepath.Join(workingDir, DefaultSignalsFile)
+	signalsJSON, err := json.MarshalIndent(signals, "", "  ")
+
+	if err != nil {
+		return err
+	}
+
+	const ownerPerms = 0644
+	if err := os.WriteFile(signalsFile, signalsJSON, ownerPerms); err != nil {
+		return fmt.Errorf("failed to write signals file %s: %w", signalsFile, err)
+	}
+
+	opts.Logger.Warnf("Written error signals to %s", signalsFile)
+
+	return nil
+}
+
+// ErrorAction represents the action to take when an error occurs
+type ErrorAction struct {
+	ShouldIgnore   bool
+	ShouldRetry    bool
+	IgnoreMessage  string
+	IgnoreSignals  map[string]interface{}
+	RetryMessage   string
+	RetryAttempts  int
+	RetrySleepSecs int
+}
+
+// ProcessError evaluates an error against the configuration and returns the appropriate action
+func (c *ErrorsConfig) ProcessError(err error, currentAttempt int) (*ErrorAction, error) {
+	if err == nil {
+		return nil, nil
+	}
+
+	errStr := err.Error()
+	action := &ErrorAction{}
+
+	// First check ignore rules
+	for _, ignoreBlock := range c.Ignore {
+		isIgnorable, err := matchesAnyPattern(errStr, ignoreBlock.IgnorableErrors)
```

Comment:
We should compile the regexes  in configurations before we call `ProcessError`, by parsing the configurations, then immediately doing the post-processing of the relevant regexes.

That would allow this method to be safer, as the user shouldn't learn that their regex is invalid in the middle of an error during a `run-all`. They should learn it while the DAG is being constructed because parsing failed.